rust/feature-flags/src/handler/flags.rs

```diff
@@ -10,29 +10,172 @@ use crate::{
+/// to detect if the request is from a client-side (browser/mobile) or server-side SDK.
+fn detect_evaluation_runtime_from_request(
+    headers: &axum::http::HeaderMap,
+    explicit_runtime: Option<EvaluationRuntime>,
+) -> Option<EvaluationRuntime> {
+    // Use explicitly passed runtime if available
+    if let Some(runtime) = explicit_runtime {
+        return Some(runtime);
+    }
+
+    // Analyze User-Agent header
+    if let Some(user_agent) = headers.get("user-agent").and_then(|v| v.to_str().ok()) {
+        // Browser patterns - these typically indicate client-side execution
+        if user_agent.contains("Mozilla/")
+            || user_agent.contains("Chrome/")
+            || user_agent.contains("Safari/")
+            || user_agent.contains("Firefox/")
+            || user_agent.contains("Edge/")
+        {
+            return Some(EvaluationRuntime::Client);
+        }
+
+        // Server SDK patterns - these indicate server-side execution
+        if user_agent.starts_with("posthog-python/")
+            || user_agent.starts_with("posthog-ruby/")
+            || user_agent.starts_with("posthog-php/")
+            || user_agent.starts_with("posthog-java/")
+            || user_agent.starts_with("posthog-go/")
+            || user_agent.starts_with("posthog-node/") // Note: server-side Node.js
+            || user_agent.starts_with("posthog-dotnet/")
+            || user_agent.starts_with("posthog-elixir/")
+            || user_agent.contains("python-requests/")
+            || user_agent.contains("curl/")
+        {
+            return Some(EvaluationRuntime::Server);
+        }
+
+        // Mobile SDK patterns - these indicate client-side mobile execution
+        if user_agent.starts_with("posthog-android/")
+            || user_agent.starts_with("posthog-ios/")
+            || user_agent.starts_with("posthog-react-native/")
+            || user_agent.starts_with("posthog-flutter/")
+        {
+            return Some(EvaluationRuntime::Client);
+        }
+    }
+
+    // Check for browser-specific headers that indicate client-side
+    if headers.contains_key("origin")
+        || headers.contains_key("referer")
+        || headers.get("sec-fetch-mode").is_some()
+        || headers.get("sec-fetch-site").is_some()
+    {
+        return Some(EvaluationRuntime::Client);
+    }
+
+    // If we can't determine, default to None (which will include flags with no runtime requirement + "all")
+    None
+}
+
+/// Filters flags to only include those that can be evaluated in the current runtime.
+/// If no runtime is specified for a flag (evaluation_runtime is None), it's included
+/// to maintain backward compatibility.
+///
+/// Note: When specific flags are requested via flag_keys, they will be evaluated
+/// after this runtime filtering step. The runtime filter helps prevent unnecessary
+/// evaluation of flags that don't match the current runtime environment.
+fn filter_flags_by_runtime(
```

Comment:
Does this PR have any implications with flags that depend on other flags? Does this mean we can't allow users to create a flag that depends on another flag with a different runtime?

Reply:
yes.  That is what it would mean. I should add some callout in the UI around this + add docs.  I'm working on the docs now; right now this feature is behind a flag and scoped to one user, but yes, that is a constraint here.